### ISSUE TITLE
python_requires is not compliant with PEP-345, PEP-566 and PEP-508

### ISSUE THREAD

**Issue by arcivanov** [pypa/setuptools#1876]
Currently `python_requires` only accepts a version specifier per PEP-440. This is an error.

Pursuant to PEP-345, as sustained in PEP-566, the specification regarding environmental markers is [as follows](https://www.python.org/dev/peps/pep-0345/#environment-markers):

```
The fields that benefit from this marker are:

Requires-Python
Requires-External
Requires-Dist
Provides-Dist
Obsoletes-Dist
Classifier
```
PEP-566 [further sustains this specification](https://www.python.org/dev/peps/pep-0566/#id16): 

```
Usage of environment markers is otherwise unchanged from PEP 345.
```
Thus, according to PEP-345, PEP-566 and PEP-508 it should be possible to specify an extremely narrow constrain such as:

`python_requires = ==3.9.1; implementation_name=='cpython' and implementation_version=='3.9.1b2'`.

As of setuptools 41.4.0 this is an error:

```
setuptools.extern.packaging.specifiers.InvalidSpecifier: Invalid specifier: '<3.9;implementation_name=='cpython''
```

**Comment by di** [comment]
Duplicate of https://github.com/pypa/setuptools/issues/1570 which was moved to https://github.com/pypa/packaging-problems/issues/209.

**Comment by arcivanov** [comment]
Not quite. The #1570 asks for something that isn't really allowed by the specs. In fact, if I understand correctly, that type of expression is not allowed for any requirement.

**Comment by pradyunsg** [comment]
pypa/packaging#224 was filed as a proposed fix for this. I've closed it, so that the discussion about what's actionable here, is kept in one place -- this issue.

**Comment by ncoghlan** [comment]
Fixing this definitely seems worthwhile to me, as it's not uncommon for projects to be affected by platform-specific standard library limitations that mean they'll need a newer base Python version in some environments. It also means that a project affected by a regression in a new feature release can clearly indicate that in their metadata:

    python_requires = >=3.8.1; python_version=='3.8'

To be fully effective, it would need to be possible to supply multiple python_requires entries, merging the entries that share the same environment marker text.

**Comment by arcivanov** [comment]
@ncoghlan 
> To be fully effective, it would need to be possible to supply multiple python_requires entries, merging the entries that share the same environment marker text.

That's certainly a noble goal to begin with, however the PEPs, as I read them do not allow it. While 

    Requirement = "name" + `SpecifierSet` + "extras" + `Marker`

the proposed change would introduce a

   SpecifierRequirement = `SpecifierSet` + `Marker`

It's entire possible I'm misreading the PEPs. Could you point me to the authority that would allow multiple `Require-Python` entries or an array of `Specifier`s with *individual* associated markers comprising a distinct `SpecifierSet`? I couldn't find any reference to such constructs in the above PEPs.

**Comment by ncoghlan** [comment]
No, you're quite right that "Requires-Python" is missing the "Multiple use" marker in PEP 345.

https://www.python.org/dev/peps/pep-0503/ also doesn't allow for multiple Requires-Python entries in its definition of the "data-requires-python" link attribute.

Environment markers aren't really useful *without* allowing multiple entries, though.

So it's probably better to treat this as a metadata spec update affecting publishing tools, the PyPI simple API, and installation tools, as otherwise it isn't likely to be very useful in practice.

**Comment by ncoghlan** [comment]
As a quick sketch of something that might work in practice:

* to create "data-requires-python", index servers combine one or more Requires-Python entries with no environment marker set using ",". This is backwards compatible with existing installers.
* if one or more Requires-Python entries with environment markers exist, index servers would compose them into a new "data-may-require-python" field.

To make this clearer, it may make sense to add a new "May-Require-Python" field that's explicitly multiple-use with environment marker support, and explicitly *disallow* environment markers in Requires-Python (since the status quo is that most tools aren't actually expecting to see environment markers there, despite what PEP 345 says).

**Comment by arcivanov** [comment]
> Environment markers aren't really useful without allowing multiple entries, though.

I'd like to point to the fact that simply prohibiting an install on a Python that is incompatible might be sufficient. Generally a marker in requirements specifies a conditional "if", but with Requires-Python the tag naturally can only mean **IFF**, as there is only one python requirement and if the condition isn't satisfied there are no alternatives.

**Comment by arcivanov** [comment]
> So it's probably better to treat this as a metadata spec update affecting publishing tools, the PyPI simple API, and installation tools, as otherwise it isn't likely to be very useful in practice.

I absolutely agree that would be something that would require setuptools and pip updates. I'm not sure about Pypi as package selection is something that is performed by the pip and  Requires-Python isn't part of package metadata exposed on the HTML page anyway.

**Comment by xavfernandez** [comment]
> Requires-Python isn't part of package metadata exposed on the HTML page anyway under PEP-503

It is (in the very same PEP you linked ;) ) - cf `data-requires-python`.

**Comment by arcivanov** [comment]
@xavfernandez 
I stand corrected, thank you! That said, PEP-503 uses "MAY" i.e. the entire system must be able to function even if the `data-requires-python` is not exposed in the anchor tag at all.

So it appears we have pypa/packaging, pypa/setuptools, pypa/pip and pypa/warehouse. The latter three will be vendoring/requiring `packaging` anyway, presumably, so creating `SpecifierRequirement` in packaging would take us 90% there.

Anything else I'm missing?

@ncoghlan @xavfernandez @pradyunsg 
There are two questions on the table:

1. Can this be accepted given that the PEP-345 already allows it, i.e. without introducing PEP-altering changes to allow multiple `Requires-Python` or novel Requirement/SpecifierRequirement expressions?

2. Are the abstractions in the proposed patch satisfactory?

**Comment by arcivanov** [comment]
@ncoghlan

> to create "data-requires-python", index servers combine one or more Requires-Python entries with no environment
> if one or more Requires-Python entries with environment markers exist

Implementation aside all this requires going through a full PEP approval process, which will take forever and requires tons of bureaucracy. 

As an alternative, you can actually already publish the same package varying only by build tags. If we bring `Requires-Python` into compliance with PEP-345, what you're describing can be achieved by build tag under PEP-427:
```
build tag

   Optional build number. Must start with a digit. A tie breaker if two wheels have the same version. Sort as the empty string if unspecified, else sort the initial digits as a number, and the remainder lexicographically.
```
If the two wheels have identical names except for build tag, but expose different `Requires-Python`, only the platform-compatible wheel will be selected for installation.

**Comment by ncoghlan** [comment]
Metadata updates stalled for a long time as we figured out what to do about metadata 2.0. PEP 566 broke that logjam by jumping to version 2.1.

Simply enabling environment markers within the existing field is going to be problematic, as adding them to PEP 503 would be backwards incompatible with existing installers.

Since the slow parts of a metadata update are ensuring backwards compatibility is covered, and the most relevant projects have been consulted, trying to avoid that isn't going to save any time, since any workaround within the existing specs would need to go through the same steps.

**Comment by arcivanov** [comment]
Very well. 

To deepen my personal understanding of your proposed change, could you please explain to me the practical usefulness of `python_requires=[">=3.4; os_name == 'posix'", ">=3.6; os_name == 'windows'"]`? 

I understand restricting installation of the module only to specific revision of CPython & OS, because of, for example, tighter coupling to a specific ABI than abi tag allows (my primary motivation for this change).

But the practical real-world utility of the above construct you proposed eludes me.

Secondly, for the uniformity sake, would you agree that having

`SpecifierRequirement` = `SpecifierSet` + `Marker` 

combined with multiple use `Requires-Python` (or whatever the new name is proposed) would be preferable to 

`SpecifierRequirement` = (`Specifier` + `Marker`)(',' `Specifier` + `Marker`)*

as the latter would require introducing additional syntactical constructs and would break with syntax in `Requirement`?

**Comment by ncoghlan** [comment]
Yes, I agree we should keep the "One specifier, one marker" principle. PEP 503 would likely need a way to combine multiple specifier requirements into a single string, but it wouldn't use a comma as the separator, and it would be specific to that index server interface.

The main purpose behind permitting multiple use is to allow people to define a baseline constraint that applies everywhere (the only currently supported functionality), and then *optionally* add conditional constraints (this would be the new part, regardless of how it is spelled in the metadata).

As an example, suppose a project uses Python's native coroutine syntax. That leads to a generic "Python-Requires: >= 3.5" constraint: no matter what platform you're on, Python 3.4 isn't going to work.

Further suppose, that when on Windows, the project uses `asyncio.WindowsProactorEventLoopPolicy` to select the proactor event loop by default. That policy object was first added in Python 3.7, so Windows users specifically would need to be on 3.7 or later for the release to work. That constraint is on the source tarball backing the release, not on any specific binary build.

Right now, there isn't a good way to express that kind of thing up front, so people don't bother, and instead report runtime errors after installation.

As a version number based example, consider a project that was affected by a regression in 3.7.0 that they couldn't work around. Right now, you have to specifically enumerate in Python-Requires all the known-incompatible versions. Thus the many Python 2/3 compatible projects with entries like "Python-Requires: >= 2.7,!=3.0",!=3.1,!=3.2,!=3.3,!=3.4".

With full multi-use and environment marker support, you'd instead be able to set minimum versions for each release series with metadata like:

    Requires-Python: >= 3.5
    Requires-Python: >=3.7.1; python_version == 3.7

The problem with that *specific* spelling is that there are a lot of tools that expect "Requires-Python" to work the way it does now, and those tools would be confused by the second entry. However, the metadata specs explicitly say to ignore fields that a tool *doesn't* recognise, thus the following spelling would be more compatible with existing versions of existing tools:

    Requires-Python: >= 3.5
    May-Require-Python: >=3.7.1; python_version == 3.7

**Comment by arcivanov** [comment]
Thank you, @ncoghlan. I understand where you're coming from. 

* What would be the desired behavior if there are multiple instances of May-Require-Python and none of them are satisfied?
* What would be the desired behavior if Requires-Python does not intersect with May-Require-Python or May-Require-Python don't intersect with itself? 
* Are individual May-Require-Python specifiers ANDed or ORed in your proposal?

I have a concern about "May-" as it sounds like MAY in RFC which is a very weak requirement.

Alternatively, maybe we could expand the use and give semantics to Requires-External since no tool so far actually obeys it functionally, Python is external to the distribution in question and the field is already multi-use and exists? In which case we would not need additional packaging abstractions and Requirement could be reused.

In which case we could define something like:

```
Requires-External: cpython >=3.4; os_name == 'posix'
Requires-External: pypy >=3.6; os_name == 'windows'
```

This would require figuring out how to ensure `pypy`, `cpython` etc are reserved and don't collide with existing names etc.

**Comment by xavfernandez** [comment]
In my opinion, and given how the main tools interpreted the existing PEP, the simplest thing to do in the short term would be to clarify/update the relevant PEPs to explain that `Requires-Python` is only expecting a `SpecifierSet`.

Future metadata updates (2.2 ?) could introduce this new concept with the `Possibly-Requires-Python` and its implications.

**Comment by arcivanov** [comment]
@xavfernandez I was under impression from @ncoghlan's explanation that PEP-566 removed the metadata out of the PEP process and delegated it to [Core Metadata Specification](https://packaging.python.org/specifications/core-metadata/) to avoid PEP process for every metadata updates.

Are you saying we should just do nothing right now? If you don't mind me asking, why?

**Comment by xavfernandez** [comment]
> I was under impression from @ncoghlan's explanation that PEP-566 removed the metadata out of the PEP process and delegated it to Core Metadata Specification to avoid PEP process for every metadata update.

As far as I know, [Core Metadata Specification](https://packaging.python.org/specifications/core-metadata/) is simply an aggregate view of the different PEP, cf [Specification Update Process](https://www.pypa.io/en/latest/specifications/#specification-update-process).

> Are you saying we should just do nothing right now? If you don't mind me asking, why?

I'm not saying we should do nothing :)
I'm suggesting that to solve the current issue (i.e. the fact that "`python_requires` is not compliant with PEP-345, PEP-566 and PEP-508") and given how the different major packaging tools (`setuptools`, `pip`, PyPI/`warehouse` and all the other tools that relied on the `packaging` library) interpreted the `Requires-Python` for the last 3 years, we should simply clarify/update the PEPs to match its defacto interpretation.

And it doesn't mean that we should not investigate new ways to specify finer metadata regarding the required python version depending on the interpreter environment but it is likely to need new fields, a new metadata version and some PEP related work.

**Comment by ncoghlan** [comment]
Right, it wouldn't require a full PEP process to remove the mention of Python-Requires from the description of environment markers, and make a note in the Python-Requires section that environment markers aren't supported in that field.

The key point allowing us to do that is that it doesn't require changes to any tools, it's just updating the spec to accurately describe the status quo.

**Comment by ncoghlan** [comment]
Back on the question of composing different requirements together, the Requires-Dist handling already covers how to do that:

* discard entries where the environment marker doesn't match the current environment
* combine all remaining entries using AND (just as if they had been part of a single comma-separated list)

Entries without an environment marker are never discarded by the first step, and thus always included by the second.

As far as a field name goes, "Env-Requires-Python" would be another plausible option, since the whole point would be to express environment specific constraints on the Python version.

(Side note: this metadata enhancement would also potentially allow projects to indicate that particular extras required a newer Python version than the basic constraint)

**Comment by arcivanov** [comment]
@ncoghlan 

In re `discard entries where the environment marker doesn't match the current environment` this can't obviously apply to `Env-Requires-Python`. If we discard an entry in `Requires-Dist` the installation proceeds without requiring a distribution. Requiring Python either through Requires-Python or Env-Requires-Python, if present, must result in failure of installation if the requirement is not satisfied, otherwise we end up with dist being installed into an environment where it can't run.

So, to me, the only logic that would make sense would be something like:

```python
# Proceed with install if no python requirements are set ...
can_install = "Requires-Python" not in metadata and \
              "Env-Requires-Python" not in metadata

if not can_install:
    # ... or Requires-Python is satisfied and 
    #  at least one Env-Requires-Python is satisfied.
    can_install = ((metadata["Requires-Python"].satisfied() 
                      if "Requires-Python" in metadata else True) 
                  and 
                  (sum(env_req_py.satisfied() 
                      for env_req_py in metadata["Env-Requires-Python"]))
                  if "Env-Requires-Python" in metadata else True))
``` 

Make sense?

**Comment by arcivanov** [comment]
> If a project wants to prevent installation in an environment, then they would need to specify something like "Env-Requires-Python: == 0.0; platform == 'windows'"

If a project wants to allow installation _only_ in specific environments then: 

```
Env-Requires-Python: ==3.7; implementation_name == 'cpython' and implementation_version=='3.7.1b2' and os_name == 'linux'
Env-Requires-Python: ==3.9; implementation_name == 'cpython' and implementation_version=='3.9.0a1' and os_name == 'windows'
```

Will install if Python is CPython that is either v 3.7.1b2 on Linux or 3.9.0a1 on Windows. Package will fail to install in all other environments and with all other Pythons.

The follow combination will fail to install anywhere due to mutually exclusive SpecifierSets:

```
Requires-Python: <=3.6
Env-Requires-Python: ==3.7; implementation_name == 'cpython' and implementation_version=='3.7.1b2' and os_name == 'linux'
Env-Requires-Python: ==3.9; implementation_name == 'cpython' and implementation_version=='3.9.0a1' and os_name == 'windows'
```

Agree?

**Comment by xavfernandez** [comment]
> > If a project wants to prevent installation in an environment, then they would need to specify something like "Env-Requires-Python: == 0.0; platform == 'windows'"
> 
> If a project wants to allow installation _only_ in specific environments then:
> 
> ```
> Env-Requires-Python: ==3.7; implementation_name == 'cpython' and implementation_version=='3.7.1b2' and os_name == 'linux'
> Env-Requires-Python: ==3.9; implementation_name == 'cpython' and implementation_version=='3.9.0a1' and os_name == 'windows'
> ```
> 
> Will install if Python is CPython that is either v 3.7.1b2 on Linux or 3.9.0a1 on Windows. Package will fail to install in all other environments and with all other Pythons.

No, that's not how environment markers read.
First you evaluate the markers (for example, `implementation_name == 'cpython' and implementation_version=='3.7.1b2' and os_name == 'linux'`):
- if the marker evaluates to `False`, the requirement `==3.7` is ignored.
- if the marker evaluates to `True`, the requirement`==3.7` is checked (and is likely to evaluate to `False` since `3.7.1b2` does not satisfy `==3.7`)

> 
> The follow combination will fail to install anywhere due to mutually exclusive SpecifierSets:
> 
> ```
> Requires-Python: <=3.6
> Env-Requires-Python: ==3.7; implementation_name == 'cpython' and implementation_version=='3.7.1b2' and os_name == 'linux'
> Env-Requires-Python: ==3.9; implementation_name == 'cpython' and implementation_version=='3.9.0a1' and os_name == 'windows'
> ```
> 
> Agree?

Again, no.
This will only fail to install on version `>3.6`.
The `Env-Requires-Python` will only be considered if `implementation_name == 'cpython' and implementation_version=='3.7.1b2' and os_name == 'linux'` or `implementation_name == 'cpython' and implementation_version=='3.9.0a1' and os_name == 'windows'` evaluate to `True` and in these cases, the `==3.7` or `==3.9` will be combined with `<=3.6` making it impossible to satisfy.

**Comment by xavfernandez** [comment]
If a project wants to allow installation only in specific environments, one could use:
```
Env-Requires-Python: ==0.0; implementation_name != 'cpython'
Env-Requires-Python: ==0.0; implementation_version != '3.7.1b2'
Env-Requires-Python: ==0.0; os_name != 'linux'
```
And you shouldn't be able to install this package on a system that does not match `implementation_name == 'cpython' and implementation_version=='3.7.1b2' and os_name == 'linux'`.

**Comment by arcivanov** [comment]
@xavfernandez 

I understand that that's how markers read in Requirement for dependencies, as in "include package in the list of required dependencies IFF marker condition is satisfied".

But we're talking about a new concept here altogether - SpecifierSets with Markers do not exist in packaging at all. So perhaps the interpretation of Markers for those specific fields can be reversed?

The purpose of environment-based dependency **requirement** and environmental **assertion** is different. Even trying to verbally describe the expression you posted above is not trivial: 

    Installing this distro will **not** succeed because it will require installing in Python 
    with a version that does **not** exist because environmental 
    condition is **not** satisfied.

We would be baking-in a _triple-negative_ logic into this, which will inevitably spill into setuptools API and documentation and every other tool under the sun. That's an unbelievably awkward construct. Imagine you're on PyPi trying to decipher the environment a particular library means?

__________

How about we reset. 

What if it's not `Env-Requires-Python`? What if it's `Env-Requires`, it consists of **only** the marker-style expression, and it's a positive assertion:

     Installation requires the following condition to be satisfied.

```
Requires-Python: >=3.7
Env-Requires: implementation_name == 'cpython' and os_name == 'linux' and implementation_version == '3.7.1b2' or os_name == 'windows'
```

In the future (and I'll be opening another ticket for this) we could have additional expressions that are related to the environment, but not necessarily a Python one, such as `cpu_march == 'nehalem' or  '0x82934ae' in cpu_features`

**Comment by xavfernandez** [comment]
> So perhaps the interpretation of Markers for those specific fields can be reversed ?

That looks like a recipe for confusion.

> Even trying to verbally describe the expression you posted above is not trivial:

It reads `Installing this distro will only succeed if the Python version matches all the requirements relevant to its environment (i.e. where the marker evaluates to True)`, which directly follows the `Installing this distro will only succeed if the Python version matches the requirement` from the current `Requires-Python` interpretation.

**Comment by arcivanov** [comment]
Very well. :)
What about the second part? :)

**Comment by ncoghlan** [comment]
We certainly *could* define a "Requires-Env" field like that if we wanted to, but that doesn't allow you to specify things like "Needs 3.7+ on Windows, 3.5+ everywhere", and doesn't fit with our general philosophy of encouraging PyPI publishers to leave their compatibility metadata relatively loose and only specify the cases that they know definitely *won't* work. Trying to specify the cases that *will* work is impossible, as there too many potential target environments.

By contrast, the "Env-Requires-Python" approach is fully consistent with the way the existing declaration fields work (it's just another kind of conditional dependency), and if someone did want to get super restrictive, they would just need to write "Env-Requires-Python: ==0.0: not (implementation_name == 'cpython' and implementation_version=='3.7.1b2' and os_name == 'linux')"

As far as pronunciation goes, an expression like "Env-Requires-Python: ==0.0; implementation_name != 'cpython'" wouldn't be read out literally, it would be pronounced as "Only CPython is supported", and "Env-Requires-Python: ==0.0; os_name != 'linux'" would be "Only Linux is supported".

We'd discourage such broad exclusions though, as they're mostly a PITA for end users.

**Comment by arcivanov** [comment]
Very well. I'm not going to push it, as my only gripe is an expression awkwardness that is rubbing me the wrong way.

Regardless, it seems that all options that have been discussed as acceptable end up with `SpecifierRequirement` which is proposed in pypa/packaging#224. Could we get this reviewed and if approved accepted in packaging right now (before Metadata 2.2 PEP) so that the prototypes can be built and PRs added to other tools? This PR doesn't change any of the existing behaviors and introduces a new abstraction we all have been discussing here at length.
